src/mgds/pipelineModules/DropTags.py

```diff
@@ -0,0 +1,155 @@
+            enabled_in_name: str,
+            probability_in_name: float,     #from 0-1, probability dropout will be applied to either each tag in list or entire section of caption after kept token
+            dropout_mode_in_name: str,      #whether to dropout all tags after kept token (FULL), randomly drop any of the tags after kept with equal probability (RANDOM), or randomly drop any tag after kept with lower probability near the start (RANDOM WEIGHTED)
+            special_tags_in_name: str,      #list of delimiter-separated tags or file path to txt/csv file which will be used as white/blacklist
+            special_tag_mode_in_name: str,  #whether the special tags are used as a WHITELIST (any tags in the list will never be dropped) or a BLACKLIST (only tags in the list can de dropped, all others kept)
+            delimiter_in_name: str,
+            regex_enabled_in_name: str,     #whether to process special tags with regex
+            keep_tags_count_in_name: str,
+            text_out_name: str,
+    ):
+        super(DropTags, self).__init__()
+        self.text_in_name = text_in_name
+        self.enabled_in_name = enabled_in_name
+        self.probability_in_name = probability_in_name
+        self.dropout_mode_in_name = dropout_mode_in_name
+        self.special_tags_in_name = special_tags_in_name
+        self.special_tag_mode_in_name = special_tag_mode_in_name
+        self.delimiter_in_name = delimiter_in_name
+        self.regex_enabled_in_name = regex_enabled_in_name
+        self.keep_tags_count_in_name = keep_tags_count_in_name
+        self.text_out_name = text_out_name
+
+    def length(self) -> int:
+        return self._get_previous_length(self.text_in_name)
+
+    def get_inputs(self) -> list[str]:
+        return [self.text_in_name, self.enabled_in_name, self.probability_in_name, self.dropout_mode_in_name, self.special_tags_in_name, self.special_tag_mode_in_name, self.delimiter_in_name, self.keep_tags_count_in_name]
+
+    def get_outputs(self) -> list[str]:
+        return [self.text_out_name]
+    
+    #convert special_tags to list depending on whether it's a newline-separated csv/txt file or a delimiter-separated string
+    #cached to reduce file read operations
+    @functools.lru_cache
+    def get_special_tags(self, sptags, delim):
+        if (sptags.endswith(".txt") and os.path.isfile(sptags)):
+            with open(sptags) as special_tags_file:
+                return [line.rstrip('\n') for line in special_tags_file]
+        elif (sptags.endswith(".csv") and os.path.isfile(sptags)):
+            with open(sptags, 'r') as special_tags_file:
+                splist = []
+                for row in csv.reader(special_tags_file):
+                    splist.append(row[0])
+                return splist
+        else:
+            return [tag.strip() for tag in sptags.split(delim)]
+
+    #parse regex expressions, create new special list based on matches   
+    def parse_regex(self, splist_in, taglist):
```

Comment:
I think you need to add some clarifying comments to this function.
1. Why does it only use regex matching if special characters are present?
2. What does the `issues caused by "\(\)" syntax` actually mean?

Reply:
1. I included that check for special characters to help speed up the function when regex processing is enabled, as I found it could slow things down a lot. Any tags in the "special list" which don't include any of those characters would only ever match to themselves (ex. "dog" would only ever match with "dog") while tags including them would potentially need to be checked (ex. "d.*g" could match "dog", "dig", "drag", "daniel carrying a bag", etc). Same with other special syntax (ex. "\d" matching to any single digit). For lists including some of both it's more efficient to only do the regex processing where needed, though I can test the execution speed more thoroughly and remove it if the effect is minimal.

2. Some booru-type tags are structured like "watercolor_(artwork)" or "raymond_(animal_crossing)", and the underscores may or may not be replaced with spaces. Most autotaggers will output them as "watercolor \(artwork\)", and using those tags during generation on most UIs also requires the same backslash to interpret the parentheses literally instead as part of (((tag weighting))). However if that is run through regex, it also uses the backslash for literal interpretation, and will interpret it as "watercolor (artwork)" and fail to match with "watercolor \(artwork\)". Those two replace lines will convert "\(" and "\)" to "\\(" and "\\)" so regex uses the first backslash to interpret the second one literally. Python also uses the same backslash for literal interpretation, so I needed to do "\\\\\(" in order to make it work. That's the only significant case I could think of where regex matching could cause unintentional problems with certain tags, but if there are others I can try to include similar replacement rules to interpret those literally as well.